### PR TITLE
ops-bedrock: Beacon-chain devnet with Dencun + Ecotone upgrade

### PR DESCRIPTION
**Pull request by sebastianst**
**Description**

This PR makes the local devnet to use a real lighthouse Beacon node.

It also enables Delta by default on all e2e tests, plus L1 enables Cancun after a few blocks, and Ecotone a shortly after.

In a follow up PR, Ecotone and Cancun will be enabled at genesis.

The PR also improves fork selection in individual tests and fixes from fork tests (e.g. one Ecotone test didn't call its activation function).

The PR introduces a new package `op-e2e/devnet` to run e2e tests against the local devnet. It provides a minimal `System` abstraction that works with the withdrawals test, which got extracted and made into a reusable test.

ℹ️ _For reviewers:_ the files in `ops-bedrock/data` are auto-generated.

**Tests**

Many fixed.

**Additional context**

Explicit fork selection in tests is somewhat improved, but is still brittle. We should completely remove fork selection by setting individual `<Fork>TimeOffset` fields in the deploy config and do it by a modular constructor of the configs using the `With...` option pattern.

Open TODOs are tracked at https://github.com/ethereum-optimism/optimism/issues/10968

**Metadata**

- Fixes https://github.com/ethereum-optimism/protocol-quest/issues/81

### CODE REVIEW

**Comment by protolambda** [comment]
Squashed and rebased on latest `develop`, in case anyone wants to run this or experiment with Ecotone locally.

**Comment by tynes** [comment]
Wondering if we plan on moving forward with this as we will be needing a devnet for interop development soon

**Comment by protolambda** [comment]
@tynes yes, moving forward with this. Currently blocked on 2 things:
- L1/L2 genesis refactor completion
- eth2-testnet-genesis tool update, to generate a L1 dencun genesis, instead of a shapella one and having to schedule an upgrade. Working on updating this (implementing dencun support in underlying libs now).

**Comment by protolambda** [comment]
`eth2-testnet-genesis` supports Dencun at genesis now

File: ops-bedrock/docker-compose.yml [hunk 1]
```diff
@@ -27,52 +29,98 @@ services:
...
       args:
         OP_STACK_GO_BUILDER: us-docker.pkg.dev/oplabs-tools-artifacts/images/op-stack-go:devnet
     image: us-docker.pkg.dev/oplabs-tools-artifacts/images/op-node:devnet
+    # The --l1.beacon.fetch-all-sidecars is required until
+    # the lighthouse image includes the following API bugfix: https://github.com/sigp/lighthouse/pull/5118
     command: >
       op-node
       --l1=ws://l1:8546
+      --l1.beacon=http://l1-bn:5052
+      --l1.beacon.fetch-all-sidecars
```
Comment (sebastianst) [inline ops-bedrock/docker-compose.yml]: we can remove `fetch-all-sidecars`

File: op-chain-ops/genesis/layer_one.go [hunk 2]
```diff
@@ -95,6 +97,17 @@ func BuildL1DeveloperGenesis(config *DeployConfig, dump *gstate.Dump, l1Deployme
...
+	beaconDepositAddr := common.HexToAddress("0x1111111111111111111111111111111111111111")
+	if err := beacondeposit.InsertEmptyBeaconDepositContract(memDB, beaconDepositAddr); err != nil {
+		return nil, fmt.Errorf("failed to insert beacon deposit contract into L1 dev genesis: %w", err)
+	}
+
+	// For 4788, make sure the 4788 beacon-roots contract is there.
+	// (required to be there before L1 Dencun activation)
+	memDB.CreateAccount(predeploys.EIP4788ContractAddr)
+	memDB.SetNonce(predeploys.EIP4788ContractAddr, 1)
+	memDB.SetCode(predeploys.EIP4788ContractAddr, predeploys.EIP4788ContractCode)
```
Comment (protolambda) [inline op-chain-ops/genesis/layer_one.go]: For consistency with actual L1 we should also set the nonce on the EIP4788 sender account, otherwise it could run in a weird failed-tx case where the destination contract address already exists.

File: op-e2e/Makefile [hunk 3]
```diff
@@ -6,7 +6,7 @@ ifdef JUNIT_FILE
 	go_test_flags = -timeout=60m -parallel=8
 else
 	go_test = go test
-	go_test_flags = -v
+	go_test_flags = # -v
```
Comment (protolambda) [inline op-e2e/Makefile]: Why this change?

**Review by protolambda — APPROVED** [review]

File: op-e2e/Makefile (hunk 3 above)
Comment (sebastianst) [reply op-e2e/Makefile]: thanks for spotting, shouldn't have been committed, artifact from local testing

File: op-chain-ops/genesis/layer_one.go (hunk 2 above)
Comment (sebastianst) [reply op-chain-ops/genesis/layer_one.go]: added, please check that it's correct

Comment (protolambda) [reply op-chain-ops/genesis/layer_one.go]: checked, matches EIP, thanks!

**Review by protolambda — APPROVED** [review]

File: op-e2e/actions/reorg_test.go [hunk 4]
```diff
@@ -148,11 +148,11 @@ func ReorgFlipFlop(gt *testing.T, deltaTimeOffset *hexutil.Uint64) {
 	verifEngClient := verifierEng.EngineClient(t, sd.RollupCfg)
 	checkVerifEngine := func() {
 		// TODO: geth preserves L2 chain with origin A1 after flip-flopping to B?
-		//ref, err := verifEngClient.L2BlockRefByLabel(t.Ctx(), eth.Unsafe)
-		//require.NoError(t, err)
-		//t.Logf("l2 unsafe head %s with origin %s", ref, ref.L1Origin)
-		//require.NotEqual(t, verifier.L2Unsafe().Hash, ref.ParentHash, "TODO off by one, engine syncs A0 after reorging back from B, while rollup node only inserts up to A0 (excl.)")
-		//require.Equal(t, verifier.L2Unsafe(), ref, "verifier safe head of engine matches rollup client")
+		// ref, err := verifEngClient.L2BlockRefByLabel(t.Ctx(), eth.Unsafe)
+		// require.NoError(t, err)
```
Comment (tynes) [inline op-e2e/actions/reorg_test.go]: This is not part of this PR but confused as to why this code is commented out in the first place

File: ops-bedrock/docker-compose.yml [hunk 5]
```diff
@@ -172,6 +217,9 @@ services:
       OP_BATCHER_BATCH_TYPE: 0
       OP_BATCHER_PLASMA_ENABLED: "${PLASMA_ENABLED}"
       OP_BATCHER_PLASMA_DA_SERVER: "http://da-server:3100"
+      # uncomment to use blobs
+      # (requires L1 Dencun and L2 Ecotone activation first)
+      # OP_BATCHER_DATA_AVAILABILITY_TYPE: blobs
```
Comment (tynes) [inline ops-bedrock/docker-compose.yml:208]: Seems like we should not have to make a diff to use this but instead use env var templating, similar to this line:

```
      OP_CHALLENGER_GAME_FACTORY_ADDRESS: ${DGF_ADDRESS}
```

so we could do something like:

```
      OP_BATCHER_DATA_AVAILABILITY_TYPE: ${OP_BATCHER_DATA_AVAILABILITY_TYPE:-calldata}
```

File: .circleci/config.yml [hunk 6]
```diff
@@ -1239,6 +1239,10 @@ jobs:
             echo '76a04354dba9980fcbc35bece2957b30 geth-alltools-linux-amd64-1.13.14-2bd6bd01.tar.gz' | md5sum -c -
             tar -xzvf geth-alltools-linux-amd64-1.13.14-2bd6bd01.tar.gz
             sudo cp geth-alltools-linux-amd64-1.13.14-2bd6bd01/* /usr/local/bin
+      - run:
+          name: Install eth2-testnet-genesis
+          command: |
+            go install -v github.com/protolambda/eth2-testnet-genesis@v0.10.0
```
Comment (tynes) [inline .circleci/config.yml]: This doesn't use `versions.json` to ensure the correct version is used

Comment (sebastianst) [reply .circleci/config.yml]: `versions.json` isn't used at all in the circleci config, see e.g. geth and nvm, which also have a hardcoded version. So I just followed this pattern. Would you still prefer to use the `versions.json` here?

File: op-e2e/actions/reorg_test.go (hunk 4 above)
Comment (sebastianst) [reply op-e2e/actions/reorg_test.go]: My auto-save-formatter created this diff. No idea what's going on here. git blame shows that @protolambda added this TODO on 2022-10-22, so it's quite dated. Would prefer not to deal with this in this PR. I'll check with him if this can be deleted.

File: ops-bedrock/docker-compose.yml (hunk 5 above)
Comment (sebastianst) [reply ops-bedrock/docker-compose.yml:208]: Good idea, I will do this in the follow up to this PR where I enable L1 Cancun at genesis in the devnet. Right now, this cannot be set to blobs because L1 Cancun only activates after 0x30 seconds.

File: op-e2e/actions/reorg_test.go (hunk 4 above)
Comment (sebastianst) [reply op-e2e/actions/reorg_test.go]: Discussed internally with proto that we can delete this TODO code, the test has the right assertions and good coverage.

File: .circleci/config.yml (hunk 6 above)
Comment (sebastianst) [reply .circleci/config.yml]: OK added this, it works. Added a follow up ticket (https://github.com/ethereum-optimism/optimism/issues/10398) to use `versions.json` more consistently in the cci config.

**Comment by tynes** [comment]
How is the progress on this coming along?

**Comment by sebastianst** [comment]
> How is the progress on this coming along?

@tynes Blocked by a bunch of other more urgent work unfortunately. Hope to get to it on Friday or Monday.

File: bedrock-devnet/devnet/__init__.py [hunk 7]
```diff
@@ -332,10 +336,10 @@ def devnet_test(paths):
...
     # And do not use devnet system addresses, to avoid breaking fee-estimation or nonce values.
     run_commands([
-        CommandPreset('erc20-test',
-          ['npx', 'hardhat',  'deposit-erc20', '--network',  'devnetL1',
-           '--l1-contracts-json-path', paths.addresses_json_path, '--signer-index', '14'],
-          cwd=paths.tasks_dir, timeout=8*60),
+        # CommandPreset('erc20-test',
+        #   ['npx', 'hardhat',  'deposit-erc20', '--network',  'devnetL1',
+        #    '--l1-contracts-json-path', paths.addresses_json_path, '--signer-index', '14'],
+        #   cwd=paths.tasks_dir, timeout=8*60)
```
Comment (sebastianst) [inline bedrock-devnet/devnet/__init__.py]: Didn't mean to commit this diff. If reviewers are happy with the op-e2e devnet test, I will just remove this altogether. Can then remove the `devnet-tasks` in a follow up.

File: op-e2e/e2eutils/secrets.go [hunk 8]
```diff
@@ -189,3 +192,14 @@ func (a *Addresses) All() []common.Address {
 		a.Mallory,
 	}
 }
+
+func (s *Secrets) AccountAtIdx(idx int) *ecdsa.PrivateKey {
```
Comment (tynes) [inline op-e2e/e2eutils/secrets.go:196]: perhaps `uint` because negative values don't make sense here?

File: bedrock-devnet/devnet/__init__.py (hunk 7 above)
Comment (tynes) [reply bedrock-devnet/devnet/__init__.py]: I would like to remove these devnet tasks, perhaps its a better idea to see them pass in CI against the new devnet before removing them

**Comment by tynes** [comment]
We are moving towards removing the L2OO and being a fault proofs only sort of system, just wanna make sure that we are on the same page and that this takes that into account, ie no deep assumptions about L2OO in this diff

File: bedrock-devnet/devnet/__init__.py (hunk 7 above)
Comment (sebastianst) [reply bedrock-devnet/devnet/__init__.py]: @tynes They don't pass and I cannot get them to work. I wasted a lot of time on this. This is the reason I switched to Go tests.

File: op-e2e/e2eutils/secrets.go (hunk 8 above)
Comment (sebastianst) [reply op-e2e/e2eutils/secrets.go:196]: It's common to use `int`s in Go for indices that don't make sense to be negative. So I just followed that convention. E.g. see stdlib, it uses `int` indices everywhere. The user of this API is usually an expert who wouldn't put in negative numbers. So I'd just leave it as is.

**Comment by sebastianst** [comment]
> We are moving towards removing the L2OO and being a fault proofs only sort of system, just wanna make sure that we are on the same page and that this takes that into account, ie no deep assumptions about L2OO in this diff

@tynes yes, this PR just adds lighthouse, the default local devnet is still FP enabled.

**Review by tynes — APPROVED** [review]

**Comment by protolambda** [comment]
Quick update about CI; I believe @sebastianst has a draft of invoking the transaction-trace on failure, to capture the `output`, and determine the error code, to fix the FP part of the withdrawal devnet tests.

Since the challenger runs in parallel to the Go test, the resolve-steps by the Go test don't always turn into success, as the claim, and then the game, may already be resolved.

**Comment by zhiqiangxu** [comment]
> This PR makes the local devnet to use a real lighthouse Beacon node.

I guess this feature is optional, right ?
